lib/plugin-api/hydrogen-kernel.js

```diff
@@ -27,6 +77,70 @@ export default class HydrogenKernel {
     }
   }
 
+  /*
+   * The language of the kernel, as specified in its kernelspec
+   */
+  get language(): string {
+    this._assertNotDestroyed();
+    return this._kernel.language;
+  }
+
+  /*
+   * The display name of the kernel, as specified in its kernelspec
+   */
+  get displayName(): string {
+    this._assertNotDestroyed();
+    return this._kernel.displayName;
+  }
+
+  /*
+   * Add a kernel middleware, which allows intercepting and issuing commands to
+   * the kernel.
+   */
+  addMiddleware(middleware: HydrogenKernelMiddleware) {
+    this._assertNotDestroyed();
+
+    const prevMiddlewareStack = this._kernel.middlewareStack;
```

Comment:
This code belongs into Kernel(=HKernel).
HydrogenKernel should be a thin wrapper around Kernel(=HKernel) that provides API V1.

Reply:
I think the comment above still stands in latest patchset. No need to duplicate the functionality in `HydrogenKernel`, `HydrogenKernelV2`, ... Just do it once in `Kernel`.